notifier/notifier.go

```diff
@@ -330,6 +319,21 @@ func (n *Manager) Run(tsets <-chan map[string][]*targetgroup.Group) {
 	}
 }
 
+// Run receives updates of target groups and triggers a reload.
+// The dispatching of notifications occurs in the background to prevent blocking the receipt of target updates.
+// Refer to https://github.com/prometheus/prometheus/issues/13676 for more details.
+func (n *Manager) Run(tsets <-chan map[string][]*targetgroup.Group) {
+	go n.sendLoop()
+	for {
+		select {
+		case <-n.ctx.Done():
+			return
```

Comment:
This has changed the behaviour of `Run`: previously, `Run` would only return once it had finished both sending notifications and processing target group updates.

However, with this change, `Run` could return before sending notifications has finished, and there's no way for callers to know when/if sending notifications has finished.

What do you think about waiting for `sendLoop()` to terminate before we return here?

Reply:
* `sending "the current (if any) batch" of notifications`.

In both cases, other notifications could still remain in the queue when `Run` returns (and that's what https://github.com/prometheus/prometheus/pull/14290 is trying to address).

Also, in both cases, the current batch (if any) will be cancelled: the request context is a child of the manager's. Previously, it was cleaner: we were sure to catch the request context cancellation and updated some metrics (even though it's useless as Prometheus is already stopping).

There is no test or doc/comments enforcing that old behavior. You can enforce that as part of your PR, or I can take care of it. I think just putting the "notifications sending" loop in the foreground should help with that.